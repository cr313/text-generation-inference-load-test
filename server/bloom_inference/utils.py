import os
import contextlib
import torch
import torch.distributed

from datetime import timedelta
from transformers.generation_logits_process import (
    LogitsProcessorList,
    TemperatureLogitsWarper,
    TopPLogitsWarper,
    TopKLogitsWarper,
)


class Sampling:
    def __call__(self, logits):
        probs = torch.nn.functional.softmax(logits, dim=-1)
        next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)
        return next_tokens


class Greedy:
    def __call__(self, logits):
        return logits.argmax(dim=-1)


class NextTokenChooser:
    def __init__(self, temperature=1.0, top_k=None, top_p=None, do_sample=False):
        warpers = LogitsProcessorList()
        # the following idea is largely copied from this PR: https://github.com/huggingface/transformers/pull/5420/files
        # all samplers can be found in `generation_utils_samplers.py`
        sampling = do_sample
        if temperature is not None and temperature != 1.0:
            temperature = float(temperature)
            warpers.append(TemperatureLogitsWarper(temperature))
            sampling = True
        if top_k is not None and top_k != 0:
            warpers.append(TopKLogitsWarper(top_k=top_k))
            sampling = True
        if top_p is not None and top_p < 1.0:
            warpers.append(TopPLogitsWarper(top_p=top_p))
            sampling = True

        self.warpers = warpers
        self.choice = Sampling() if sampling else Greedy()

    def __call__(self, input_ids, scores):
        scores = self.warpers(input_ids, scores)
        next_ids = self.choice(scores)
        return next_ids.unsqueeze(-1)


class StoppingCriteria:
    def __init__(self, max_new_tokens=20):
        self.max_new_tokens = max_new_tokens
        self.current_tokens = 0

    def __call__(self, all_ids):
        self.current_tokens += 1
        if self.current_tokens >= self.max_new_tokens:
            return True
        return False


def initialize_torch_distributed():
    rank = int(os.getenv("RANK", "0"))
    world_size = int(os.getenv("WORLD_SIZE", "1"))

    if torch.cuda.is_available():
        # initialized `torch.distributed`
        # Set the device id.
        assert world_size <= torch.cuda.device_count(), "Each process is one gpu"
        device = rank % torch.cuda.device_count()
        torch.cuda.set_device(device)
        backend = "nccl"
    else:
        backend = "gloo"

    # Call the init process.
    torch.distributed.init_process_group(
        backend=backend,
        world_size=world_size,
        rank=rank,
        timeout=timedelta(seconds=60),
        init_method="tcp://localhost:6000",
    )

    return torch.distributed.distributed_c10d._get_default_group(), rank, world_size


@contextlib.contextmanager
def set_default_dtype(dtype):
    saved_dtype = torch.get_default_dtype()
    torch.set_default_dtype(dtype)
    try:
        yield
    finally:
        torch.set_default_dtype(saved_dtype)
