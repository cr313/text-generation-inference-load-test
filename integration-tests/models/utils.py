import time

from aiohttp import ClientConnectorError, ClientOSError, ServerDisconnectedError
from text_generation import AsyncClient


async def health_check(client: AsyncClient, timeout: int = 60):
    assert timeout > 0
    for _ in range(timeout):
        try:
            await client.generate("test")
            return
        except (ClientConnectorError, ClientOSError, ServerDisconnectedError) as e:
            time.sleep(1)
    raise RuntimeError("Health check failed")
