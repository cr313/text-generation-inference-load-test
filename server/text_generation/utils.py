import os
import torch
import torch.distributed

from datetime import timedelta

from functools import partial
from joblib import Parallel, delayed
from huggingface_hub import HfApi, hf_hub_download, try_to_load_from_cache
from huggingface_hub.utils import LocalEntryNotFoundError
from tqdm import tqdm
from transformers.generation_logits_process import (
    LogitsProcessorList,
    TemperatureLogitsWarper,
    TopPLogitsWarper,
    TopKLogitsWarper,
)


class Sampling:
    def __call__(self, logits):
        probs = torch.nn.functional.softmax(logits, dim=-1)
        next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)
        return next_tokens


class Greedy:
    def __call__(self, logits):
        return logits.argmax(dim=-1)


class NextTokenChooser:
    def __init__(self, temperature=1.0, top_k=None, top_p=None, do_sample=False):
        warpers = LogitsProcessorList()
        # the following idea is largely copied from this PR: https://github.com/huggingface/transformers/pull/5420/files
        # all samplers can be found in `generation_utils_samplers.py`
        sampling = do_sample
        if temperature is not None and temperature != 1.0:
            temperature = float(temperature)
            warpers.append(TemperatureLogitsWarper(temperature))
            sampling = True
        if top_k is not None and top_k != 0:
            warpers.append(TopKLogitsWarper(top_k=top_k))
            sampling = True
        if top_p is not None and top_p < 1.0:
            warpers.append(TopPLogitsWarper(top_p=top_p))
            sampling = True

        self.warpers = warpers
        self.choice = Sampling() if sampling else Greedy()

    def __call__(self, input_ids, scores):
        scores = self.warpers(input_ids, scores)
        next_ids = self.choice(scores)
        return next_ids.unsqueeze(-1)


class StoppingCriteria:
    def __init__(self, max_new_tokens=20):
        self.max_new_tokens = max_new_tokens
        self.current_tokens = 0

    def __call__(self, all_ids):
        self.current_tokens += 1
        if self.current_tokens >= self.max_new_tokens:
            return True
        return False


def initialize_torch_distributed():
    rank = int(os.getenv("RANK", "0"))
    world_size = int(os.getenv("WORLD_SIZE", "1"))

    if torch.cuda.is_available():
        # initialized `torch.distributed`
        # Set the device id.
        assert world_size <= torch.cuda.device_count(), "Each process is one gpu"
        device = rank % torch.cuda.device_count()
        torch.cuda.set_device(device)
        backend = "nccl"
    else:
        backend = "gloo"

    # Call the init process.
    torch.distributed.init_process_group(
        backend=backend,
        world_size=world_size,
        rank=rank,
        timeout=timedelta(seconds=60),
    )

    return torch.distributed.distributed_c10d._get_default_group(), rank, world_size


def weight_hub_files(model_name, extension=".safetensors"):
    """Get the safetensors filenames on the hub"""
    api = HfApi()
    info = api.model_info(model_name)
    filenames = [s.rfilename for s in info.siblings if s.rfilename.endswith(extension)]
    return filenames


def weight_files(model_name, extension=".safetensors"):
    """Get the local safetensors filenames"""
    filenames = weight_hub_files(model_name, extension)
    files = []
    for filename in filenames:
        cache_file = try_to_load_from_cache(model_name, filename=filename)
        if cache_file is None:
            raise LocalEntryNotFoundError(
                f"File {filename} of model {model_name} not found in "
                f"{os.getenv('HUGGINGFACE_HUB_CACHE', 'the local cache')}. "
                f"Please run `text-generation-server download-weights {model_name}` first."
            )
        files.append(cache_file)

    return files


def download_weights(model_name, extension=".safetensors"):
    """Download the safetensors files from the hub"""
    filenames = weight_hub_files(model_name, extension)

    download_function = partial(
        hf_hub_download, repo_id=model_name, local_files_only=False
    )
    # FIXME: fix the overlapping progress bars
    files = Parallel(n_jobs=5)(
        delayed(download_function)(filename=filename) for filename in tqdm(filenames)
    )
    return files
