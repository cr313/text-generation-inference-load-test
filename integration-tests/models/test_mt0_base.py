import pytest

from utils import health_check


@pytest.fixture(scope="module")
def mt0_base(launcher):
    with launcher("bigscience/mt0-base") as client:
        yield client


@pytest.mark.asyncio
async def test_mt0_base(mt0_base, snapshot_test):
    await health_check(mt0_base, 60)

    response = await mt0_base.generate(
        "Why is the sky blue?",
        max_new_tokens=10,
        top_p=0.9,
        seed=0,
    )

    assert response.details.generated_tokens == 5
    assert snapshot_test(response)


@pytest.mark.asyncio
async def test_mt0_base_all_params(mt0_base, snapshot_test):
    await health_check(mt0_base, 60)

    response = await mt0_base.generate(
        "Why is the sky blue?",
        max_new_tokens=10,
        repetition_penalty=1.2,
        return_full_text=True,
        stop_sequences=["test"],
        temperature=0.5,
        top_p=0.9,
        top_k=10,
        truncate=5,
        typical_p=0.9,
        watermark=True,
        seed=0,
    )

    assert response.details.generated_tokens == 10
    assert snapshot_test(response)


@pytest.mark.asyncio
async def test_mt0_base_load(mt0_base, generate_load, snapshot_test):
    await health_check(mt0_base, 60)

    responses = await generate_load(
        mt0_base,
        "Why is the sky blue?",
        max_new_tokens=10,
        n=4,
    )

    assert len(responses) == 4

    assert snapshot_test(responses)
