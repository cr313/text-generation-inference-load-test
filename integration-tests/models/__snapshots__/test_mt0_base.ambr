# serializer version: 1
# name: test_mt0_base
  dict({
    'details': dict({
      'best_of_sequences': None,
      'finish_reason': <FinishReason.EndOfSequenceToken: 'eos_token'>,
      'generated_tokens': 5,
      'prefill': list([
        dict({
          'id': 0,
          'text': '<pad>',
        }),
      ]),
      'seed': 0,
      'tokens': list([
        dict({
          'id': 926,
          'special': False,
          'text': 'To',
        }),
        dict({
          'id': 18295,
          'special': False,
          'text': ' sell',
        }),
        dict({
          'id': 7868,
          'special': False,
          'text': ' things',
        }),
        dict({
          'id': 260,
          'special': False,
          'text': '.',
        }),
        dict({
          'id': 1,
          'special': True,
          'text': '</s>',
        }),
      ]),
    }),
    'generated_text': 'To sell things.',
  })
# ---
# name: test_mt0_base_all_params
  dict({
    'details': dict({
      'best_of_sequences': None,
      'finish_reason': <FinishReason.Length: 'length'>,
      'generated_tokens': 10,
      'prefill': list([
        dict({
          'id': 0,
          'text': '<pad>',
        }),
      ]),
      'seed': 0,
      'tokens': list([
        dict({
          'id': 16017,
          'special': False,
          'text': 'blue',
        }),
        dict({
          'id': 20495,
          'special': False,
          'text': ' sky',
        }),
        dict({
          'id': 259,
          'special': False,
          'text': ' ',
        }),
        dict({
          'id': 15484,
          'special': False,
          'text': 'appear',
        }),
        dict({
          'id': 345,
          'special': False,
          'text': 'ed',
        }),
        dict({
          'id': 288,
          'special': False,
          'text': ' to',
        }),
        dict({
          'id': 35622,
          'special': False,
          'text': ' cloud',
        }),
        dict({
          'id': 263,
          'special': False,
          'text': 's',
        }),
        dict({
          'id': 14701,
          'special': False,
          'text': ' above',
        }),
        dict({
          'id': 751,
          'special': False,
          'text': ' all',
        }),
      ]),
    }),
    'generated_text': 'Why is the sky blue?blue sky appeared to clouds above all',
  })
# ---
# name: test_mt0_base_load
  list([
    dict({
      'details': dict({
        'best_of_sequences': None,
        'finish_reason': <FinishReason.EndOfSequenceToken: 'eos_token'>,
        'generated_tokens': 6,
        'prefill': list([
          dict({
            'id': 0,
            'text': '<pad>',
          }),
        ]),
        'seed': None,
        'tokens': list([
          dict({
            'id': 259,
            'special': False,
            'text': '',
          }),
          dict({
            'id': 39261,
            'special': False,
            'text': 'Because',
          }),
          dict({
            'id': 609,
            'special': False,
            'text': ' it',
          }),
          dict({
            'id': 339,
            'special': False,
            'text': ' is',
          }),
          dict({
            'id': 16017,
            'special': False,
            'text': ' blue',
          }),
          dict({
            'id': 1,
            'special': True,
            'text': '</s>',
          }),
        ]),
      }),
      'generated_text': 'Because it is blue',
    }),
    dict({
      'details': dict({
        'best_of_sequences': None,
        'finish_reason': <FinishReason.EndOfSequenceToken: 'eos_token'>,
        'generated_tokens': 6,
        'prefill': list([
          dict({
            'id': 0,
            'text': '<pad>',
          }),
        ]),
        'seed': None,
        'tokens': list([
          dict({
            'id': 259,
            'special': False,
            'text': '',
          }),
          dict({
            'id': 39261,
            'special': False,
            'text': 'Because',
          }),
          dict({
            'id': 609,
            'special': False,
            'text': ' it',
          }),
          dict({
            'id': 339,
            'special': False,
            'text': ' is',
          }),
          dict({
            'id': 16017,
            'special': False,
            'text': ' blue',
          }),
          dict({
            'id': 1,
            'special': True,
            'text': '</s>',
          }),
        ]),
      }),
      'generated_text': 'Because it is blue',
    }),
    dict({
      'details': dict({
        'best_of_sequences': None,
        'finish_reason': <FinishReason.EndOfSequenceToken: 'eos_token'>,
        'generated_tokens': 6,
        'prefill': list([
          dict({
            'id': 0,
            'text': '<pad>',
          }),
        ]),
        'seed': None,
        'tokens': list([
          dict({
            'id': 259,
            'special': False,
            'text': '',
          }),
          dict({
            'id': 39261,
            'special': False,
            'text': 'Because',
          }),
          dict({
            'id': 609,
            'special': False,
            'text': ' it',
          }),
          dict({
            'id': 339,
            'special': False,
            'text': ' is',
          }),
          dict({
            'id': 16017,
            'special': False,
            'text': ' blue',
          }),
          dict({
            'id': 1,
            'special': True,
            'text': '</s>',
          }),
        ]),
      }),
      'generated_text': 'Because it is blue',
    }),
    dict({
      'details': dict({
        'best_of_sequences': None,
        'finish_reason': <FinishReason.EndOfSequenceToken: 'eos_token'>,
        'generated_tokens': 6,
        'prefill': list([
          dict({
            'id': 0,
            'text': '<pad>',
          }),
        ]),
        'seed': None,
        'tokens': list([
          dict({
            'id': 259,
            'special': False,
            'text': '',
          }),
          dict({
            'id': 39261,
            'special': False,
            'text': 'Because',
          }),
          dict({
            'id': 609,
            'special': False,
            'text': ' it',
          }),
          dict({
            'id': 339,
            'special': False,
            'text': ' is',
          }),
          dict({
            'id': 16017,
            'special': False,
            'text': ' blue',
          }),
          dict({
            'id': 1,
            'special': True,
            'text': '</s>',
          }),
        ]),
      }),
      'generated_text': 'Because it is blue',
    }),
  ])
# ---
